### PR TITLE
C#: Tool to generate stubs for qltests

### PR DESCRIPTION
**Pull request by calumgrant**
This is a tool used to generate a file called `stubs.cs` in a qltest directory. `stubs.cs` contains a minimal set of declarations that enable the test to compile without third-party dependencies, making the test self-contained.

The way this works is via a Python script, `make_stubs.py`, that 

1. Performs a regular qltest, using `--leave-temp-files` to create a database of the test.
2. Runs a query `MinimalStubsFromSource.ql` on that database, to figure out which symbols were needed by the test, then uses QL to generate the C# source code.
3. Re-runs the test, with the new `stubs.cs` in place, to check that the generated code compiles and runs.

The overall workflow is as follows:

1. Write the qltest as normal, and reference 3rd-party assemblies via `semmle-extractor-options` comments in the source code as before.
2. Run `make_stubs.py` to generate `stubs.cs`
3. If test no longer compiles, fix up `stubs.cs` so that the test compiles and passes. Unfortunately, the code generation can be a bit buggy.
4. Edit the `semmle-extractor-options` line to remove the third-party dependencies.
5. Re-run the qltest to check that it works.
6. Commit `stubs.cs` and other changes.

See for example https://github.com/denislevin/ql/pull/2 where I have done this.

Are these the right locations for these files and tests?

This approach was an experiment but it seems to work. I still think it would be nicer to automate this process further, but this PR works and solves the immediate problem.

@lukecartey @pavgust @hvitved

### CODE REVIEW

File: csharp/ql/src/Stubs/make_stubs.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    print("Test directory does not contain .ql files. Please specify a working qltest directory.")
+    exit(1)
+
+if not foundCS:
+    print("Test directory does not contain .cs files. Please specify a working qltest directory.")
+    exit(1)
+
+cmd = 'odasa selfTest'
+print(cmd)
+if os.system(cmd):
```
Comment (hvitved) [inline csharp/ql/src/Stubs/make_stubs.py]: I would normally use `subprocess.check_call` instead.

File: csharp/ql/src/Stubs/Stubs.qll [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+  }
+
+  private string stubBaseTypesString() {
+    if this instanceof Enum then
+      result = ""
+    else if exists(getAnInterestingBaseType()) then
+      result = " : " +
+        concat(int i, ValueOrRefType t |
+          t = this.getAnInterestingBaseType() and (if t instanceof Class then i=0 else i=1) |
+          stubClassName(t), ", " order by i
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/ql/src/Stubs/Stubs.qll]: Yes.

File: csharp/ql/src/Stubs/Stubs.qll [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+  }
+
+  final int getChildNamespaceCount() {
+    result = count(GeneratedNamespace g | g.getParentNamespace() = this)
+  }
+
+  private string getSubNamespaces(int n) {
+    result = getChildNamespace(n).getStubs() + getSubNamespaces(n+1)
+    or
+    n = getChildNamespaceCount() and result = ""
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/ql/src/Stubs/Stubs.qll]: That worked, great. Initially I was fighting some negative recursion but it seems to have been fixed.

File: csharp/ql/src/Stubs/Stubs.qll [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+}
+
+private string stubExplicitImplementation(Member c) {
+  if exists(c.(Virtualizable).getExplicitlyImplementedInterface())
+  then result = stubClassName(c.(Virtualizable).getExplicitlyImplementedInterface()) + "."
+  else result = ""
+}
+
+private string stubMember(Member m) {
+  exists(Method c | m = c and not m.getDeclaringType() instanceof Enum |
```
[the remark this answers is not in the document]
Comment (calumgrant) [reply csharp/ql/src/Stubs/Stubs.qll]: In this particular case, I think it makes more sense to reuse the `c` variable etc,

File: csharp/ql/src/Stubs/make_stubs.py (hunk 1 above)
Comment (calumgrant) [reply csharp/ql/src/Stubs/make_stubs.py]: That's much better, thanks.

**Pull request merged**